src/3d/qgsgoochmaterialsettings.cpp

```diff
@@ -0,0 +1,115 @@
+/***************************************************************************
+  qgsgoochmaterialsettings.cpp
+  --------------------------------------
+  Date                 : July 2020
+  Copyright            : (C) 2020 by Nyall Dawson
+  Email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgoochmaterialsettings.h"
+
+#include "qgssymbollayerutils.h"
+#include "qgslinematerial_p.h"
+#include <Qt3DExtras/QGoochMaterial>
+
+QString QgsGoochMaterialSettings::type() const
+{
+  return QStringLiteral( "gooch" );
+}
+
+QgsAbstractMaterialSettings *QgsGoochMaterialSettings::create()
+{
+  return new QgsGoochMaterialSettings();
+}
+
+bool QgsGoochMaterialSettings::supportsTechnique( QgsMaterialSettingsRenderingTechnique technique )
+{
+  switch ( technique )
+  {
+    case QgsMaterialSettingsRenderingTechnique::Triangles:
+    case QgsMaterialSettingsRenderingTechnique::Terrain:
+      return true;
+
+    case QgsMaterialSettingsRenderingTechnique::Lines:
+    case QgsMaterialSettingsRenderingTechnique::InstancedPoints:
+    case QgsMaterialSettingsRenderingTechnique::Points:
+      return false;
+  }
+  return false;
+}
+
+QgsGoochMaterialSettings *QgsGoochMaterialSettings::clone() const
+{
+  return new QgsGoochMaterialSettings( *this );
+}
+
+void QgsGoochMaterialSettings::readXml( const QDomElement &elem, const QgsReadWriteContext & )
+{
+  mWarm = QgsSymbolLayerUtils::decodeColor( elem.attribute( QStringLiteral( "warm" ), QStringLiteral( "107,0,107" ) ) );
+  mCool = QgsSymbolLayerUtils::decodeColor( elem.attribute( QStringLiteral( "cool" ), QStringLiteral( "255,130,0" ) ) );
+  mDiffuse = QgsSymbolLayerUtils::decodeColor( elem.attribute( QStringLiteral( "diffuse" ), QStringLiteral( "178,178,178" ) ) );
+  mSpecular = QgsSymbolLayerUtils::decodeColor( elem.attribute( QStringLiteral( "specular" ) ) );
+  mShininess = elem.attribute( QStringLiteral( "shininess2" ), QStringLiteral( "100" ) ).toFloat();
```

Comment:
Just curious... why is it shininess2 ? :-)

Reply:
I can't work out the behaviour of the shininess parameter in the Qt materials, but it seems to be very different with the new material classes vs the older phong material class. The values which are reasonable for QPhongMaterial give silly results for gooch material. And then when switching between material types, we attempt to read the older material xml using the new material, in order to restore as many compatible properties as possible (e.g. so that the diffuse color doesn't get reset when you flip between material types). So this was a hack to avoid copying the invalid phong shininess values across the gooch, and vice-versa...